packages/react-query-next-experimental/src/HydrationStreamProvider.tsx

```diff
@@ -0,0 +1,184 @@
+   *   - We check if `window[id]` is set to an array and call `push()` on all the entries which will call `onEntries()` with the new entries
+   *   - We replace `window[id]` with a `push()`-method that will be called whenever new entries are received
+   **/
+  function UseClientHydrationStreamProvider(props: {
+    children: React.ReactNode
+    /**
+     * Optional transformer to serialize/deserialize the data
+     * Example devalue, superjson et al
+     */
+    transformer?: DataTransformer
+    /**
+     * **Client method**
+     * Called in the browser when new entries are received
+     */
+    onEntries: (entries: TShape[]) => void
+    /**
+     * **Server method**
+     * onFlush is called on the server when the cache is flushed
+     */
+    onFlush?: () => TShape[]
+  }) {
+    // unique id for the cache provider
+    const id = `__RQ${React.useId()}`
+    const idJSON = JSON.stringify(id)
+
+    const [transformer] = React.useState(
+      () =>
+        (props.transformer ?? {
+          // noop
+          serialize: (obj: any) => obj,
+          deserialize: (obj: any) => obj,
+        }) as unknown as TypedDataTransformer<TShape>,
+    )
+
+    // <server stuff>
+    const [stream] = React.useState<TShape[]>(() => {
+      if (typeof window !== 'undefined') {
+        return {
+          push() {
+            // no-op on the client
+          },
+        } as unknown as TShape[]
+      }
+      return []
+    })
+    const count = React.useRef(0)
+    useServerInsertedHTML(() => {
+      // This only happens on the server
+      stream.push(...(props.onFlush?.() ?? []))
+
+      if (!stream.length) {
+        return null
+      }
+      // console.log(`pushing ${stream.length} entries`)
+      const serializedCacheArgs = stream
+        .map((entry) => transformer.serialize(entry))
+        .map((entry) => JSON.stringify(entry))
+        .join(',')
+
+      // Flush stream
+      stream.length = 0
+
+      const html: string[] = [
+        `window[${idJSON}] = window[${idJSON}] || [];`,
+        `window[${idJSON}].push(${serializedCacheArgs});`,
+      ]
+      return (
+        <script
+          key={count.current++}
+          dangerouslySetInnerHTML={{
+            __html: html.join(''),
+          }}
+        />
+      )
+    })
+    // </server stuff>
+
+    // <client stuff>
+    const onEntriesRef = React.useRef(props.onEntries)
+    React.useEffect(() => {
+      onEntriesRef.current = props.onEntries
+    })
+
+    React.useEffect(() => {
+      // Client: consume cache:
+      const onEntries = (...serializedEntries: Serialized<TShape>[]) => {
+        const entries = serializedEntries.map((serialized) =>
+          transformer.deserialize(serialized),
+        )
+        onEntriesRef.current(entries)
+      }
+
+      const win = window as any
+      // Register cache consumer
+      const winStream: Array<Serialized<TShape>> = win[id] ?? []
+
+      onEntries(...winStream)
```

Comment:
This is the only time anything hydrates, right? Can we always do that in an effect? Is the effect for an entry always guaranteed to run before React hydrates the corresponding DOM? Imagine for example if this provider and a section that has a query flushes to the stream at the same time, will the effect have a chance to run before that section is hydrated by React then? If not, there will be hydration mismatches.

Reply:
> This is the only time anything hydrates, right? 

Yes

> Can we always do that in an effect? Is the effect for an entry always guaranteed to run before React hydrates the corresponding DOM? Imagine for example if this provider and a section that has a query flushes to the stream at the same time, will the effect have a chance to run before that section is hydrated by React then? If not, there will be hydration mismatches.


I am not entirely sure to be honest. All of this stuff is experimental

Reply:
> I am not entirely sure to be honest. All of this stuff is experimental

Me neither and oh yeah, no worries, that's why I'm trying to think this through. 😃

Okay, so I looked at this closer and realized that it's only the first time this happens in an effect, after that, `window[id].push()` calls it directly when the script is evaluated which should always be _before_ hydrating that Suspense boundary.

So we could still theoretically have the problem that this effect hasn't run before the first Suspense boundary hydrates right (unless effects for previous boundaries _are_ guaranteed to run before hydration of child boundary)?

But at least, after that, right now the hydration happens before hydrating the Suspense boundary that is streaming in. Since this only happens on SSR/first page load, it should be pretty unlikely that we stream in a query that is already in the cache, at least with some NEW data. Even if we do, I think it's fine in this case (but wont be for the `<HydrateBoundary>` case since that will run on page navigations too).